pkg/tar/tar.go

```diff
@@ -95,124 +74,80 @@ type Reader interface {
-func (t *stiTar) StreamDirAsTarWithCallback(source string, writer io.Writer, walkFn filepath.WalkFunc, modifyInplace bool) error {
-	f, err := os.Open(source)
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-	info, err := f.Stat()
-	if err != nil {
-		return err
-	}
-	if !info.IsDir() {
-		return fmt.Errorf("the source %q has to be directory, not a file", source)
-	}
-	destDir := source
-	if !modifyInplace {
-		fs := util.NewFileSystem()
-		tmpDir, err := ioutil.TempDir("", "s2i-")
-		if err != nil {
-			return err
-		}
-		defer os.RemoveAll(tmpDir)
-		if err = fs.Copy(source, tmpDir); err != nil {
-			return err
-		}
-		destDir = tmpDir
-	}
-	if err := filepath.Walk(destDir, walkFn); err != nil {
-		return err
-	}
-	return t.CreateTarStream(destDir, false, writer)
+// RenameAdapter renames files and directories inline as a tarfile is being
+// written
+type RenameAdapter struct {
+	Writer
+	Old string
+	New string
 }
 
-// StreamFileAsTar streams the source file as a tar archive.
-// The permissions of all files in archive is changed to 0666.
-func (t *stiTar) StreamFileAsTar(source, name string, writer io.Writer) error {
-	makeFileWorldWritable := func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		// Skip chmod if on windows OS and for symlinks
-		if runtime.GOOS == "windows" || info.Mode()&os.ModeSymlink != 0 {
-			return nil
-		}
-		return os.Chmod(path, 0666)
+// WriteHeader renames files and directories inline as a tarfile is being
+// written
+func (a RenameAdapter) WriteHeader(hdr *tar.Header) error {
+	if hdr.Name == a.Old {
+		hdr.Name = a.New
+	} else if strings.HasPrefix(hdr.Name, a.Old+"/") {
+		hdr.Name = a.New + hdr.Name[len(a.Old):]
 	}
-	return t.StreamFileAsTarWithCallback(source, name, writer, makeFileWorldWritable, false)
+
+	return a.Writer.WriteHeader(hdr)
 }
 
-// StreamFileAsTarWithCallback streams the source file as a tar archive.
-func (t *stiTar) StreamFileAsTarWithCallback(source, name string, writer io.Writer, walkFn filepath.WalkFunc, modifyInplace bool) error {
-	f, err := os.Open(source)
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-	info, err := f.Stat()
-	if err != nil {
-		return err
-	}
-	if info.IsDir() {
-		return fmt.Errorf("the source %q has to be regular file, not directory", source)
-	}
-	fs := util.NewFileSystem()
-	tmpDir, err := ioutil.TempDir("", "s2i-")
-	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(tmpDir)
-	dst := filepath.Join(tmpDir, name)
-	if err := fs.Copy(source, dst); err != nil {
-		return err
-	}
-	fileInfo, fileErr := os.Stat(dst)
-	if err := walkFn(dst, fileInfo, fileErr); err != nil {
-		return err
+// New creates a new Tar
+func New(fs util.FileSystem) Tar {
+	return &stiTar{
+		FileSystem: fs,
+		exclude:    DefaultExclusionPattern,
+		timeout:    defaultTimeout,
 	}
-	return t.CreateTarStream(tmpDir, false, writer)
+}
+
+// stiTar is an implementation of the Tar interface
+type stiTar struct {
```

Comment:
@jim-minter ^^

\<fun\>
will we ever get rid of the sti's in the code base... ;)
\</fun\>